File: backend/feed-service/src/main/java/ru/maxigram/backend/feedservice/service/LikeService.java
package ru.maxigram.backend.feedservice.service;

import java.util.Collection;
import ru.maxigram.backend.feedservice.entity.Comment;
import ru.maxigram.backend.feedservice.entity.Like;
import ru.maxigram.backend.feedservice.entity.Post;

/** Сервис для работы с лайками. */
public interface LikeService {

  /**
   * Ставит лайк на пост.
   *
   * @param post пост
   * @param authorId идентификатор автора
   * @return поставленный лайк
   */
  Like likePost(Post post, String authorId);

  /**
   * Ставит лайк на комментарий.
   *
   * @param comment комментарий
   * @param authorId идентификатор автора
   * @return поставленный лайк
   */
  Like likeComment(Comment comment, String authorId);

  /**
   * Возвращает идентификаторы всех лайкнутых пользователем постов.
   *
   * @param userId идентификатор пользователя
   * @return идентификаторы всех лайкнутых пользователем постов
   */
  Collection<Long> getPostIdsLikedByUser(String userId);

  /**
   * Возвращает идентификаторы всех лайкнутых пользователем комментариев.
   *
   * @param userId идентификатор пользователя
   * @return идентификаторы всех лайкнутых пользователем комментариев
   */
  Collection<Long> getCommentIdsLikedByUser(String userId);
}

File: src/main/kotlin/org/maxizenit/maxigram/feed/LikeService.kt
package org.maxizenit.maxigram.feed

import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class LikeService(
    private val likes: LikeRepository,
    private val posts: PostRepository,
    private val comments: CommentRepository,
    private val events: ApplicationEventPublisher,
) {

    /** Idempotent (onConflictDoNothing); v1 lost likes entirely due to JPA cascade mistakes. */
    fun likePost(postId: Long, userId: UUID) {
        val authorId = posts.authorOf(postId) ?: throw PostNotFoundException(postId)
        if (likes.likePost(postId, userId)) {
            events.publishEvent(PostLiked(postId, authorId, userId))
        }
    }

    fun unlikePost(postId: Long, userId: UUID) = likes.unlikePost(postId, userId)

    fun likeComment(commentId: Long, userId: UUID) {
        if (!comments.existsById(commentId)) throw CommentNotFoundException(commentId)
        likes.likeComment(commentId, userId)
    }

    fun unlikeComment(commentId: Long, userId: UUID) = likes.unlikeComment(commentId, userId)
}
